### ISSUE TITLE
Getting HomeKit to manipulate more than 3 lights at a time often fails to manipulate every light correctly

### ISSUE THREAD

**Issue by kavehv** [nfarina/homebridge#2007]
I'm using homebridge-isy-js to control lights controlled by my ISY 994 controller.  HomeKit generally doesn't have any issue except when asking it to manipulate more than 3 at a time (with Siri or in a Scene).  I filed a bug with Apple, they seem to think the fault is with HomeBridge

Try turning on a set of 4 or 5 lights with Siri or Home

### Expected Behavior
Everything responds as you'd expect

### Current Behavior
A random set of lights will actually respond

**Comment by garrydwilms** [comment]
I have the same issue. Can you show examples of how you combined set events?

**Comment by ebaauw** [comment]
See https://github.com/ebaauw/homebridge-hue/blob/84fa48e4ccbeccf364a01d02612151832f1f8ea3/lib/HueLight.js#L1192-L1238

I call `this.request()` from the `set` event listeners.  This collects the new state attributes in `this.desiredState`, sets a timeout for `this.put()` (if there isn't one already), and returns a new promise.  `this.put()` fires the actual request and resolves the `this.request()` promises.  The timeout is configurable from config.json, default 20 msec.

For throttling, I simply use the `deferred` package, see https://github.com/ebaauw/homebridge-hue/blob/84fa48e4ccbeccf364a01d02612151832f1f8ea3/lib/HueBridge.js#L224-L227.  Again, the number of parallel requests is configurable through config.json.  The default depends on the Hue bridge model (the older version is far less powerful).

**Comment by NorthernMan54** [comment]
@garrydwilms Sorry, I haven't delved into the details, but is the issue possibly caused by multiple requests being executed in parallel?  And the device/controller stumbling over multiple requests?

I had something similar with my X-10 controller plugin ( heyu ), and I had someone submit a pull request with a simple queuing approach, that caused all requests to queue before execution.  Not as sophisticated to what Erik has, but a simple implementation.

https://github.com/NorthernMan54/homebridge-heyu/commit/f6e654fab6af6f84e0f635ebeb01b94f8a5b36d3
